### PR TITLE
1600 move away from bower direct

### PR DESCRIPTION
**Pull request by d3sandoval**
# Addressing #1600 (move away from bower)

## Major changes
- `bower.json` has been deprecated in favor of `package.json` - where all the front-end assets are installed (unless otherwise installed in composer or zend_modules)
- some packages could not be found on npm or have multiple versions - napa is used to install these packages
- some packages do not have distributable versions for the version openemr requires - these have been added/kept in `public/assets/manual-added-packages` or, if modified, `publc/assets/modified`. All other vendor files are gitignored.
- gulp is what copies all of the installed files from node_modules into the public/assets directory. It has some rules to try to remove unnecessary files. The `gulp` step runs after `npm install` via the `post install` step.
- The latest version of each library in use by openemr is now canonically named (e.g. `jquery-3-1-1` is now just `jquery` in the `public/assets` directory). All `script` and `link` tags have been changed to point to these generic versions instead of explicit versions.
    - All the remaining explicit versions are "out of date" versions of libraries. These will probably need a separate issue called [Upgrade or replace outdated front-end packages]() <-- add link here.

### Minor changes
- Added an `npm install` to the build step. As noted in the TODOs, this would make more sense as postinstall step in composer - since it's vendor related.
- reordered the config.yaml and added entries for a few of the dependencies. The original intent was to use this function to migrate inline `<script>` tags into `Header::setupHeader()` statements. This was not done in this PR because it was taking too long and was difficult to test what broke what. You can see the attempt at these changes [in this branch](https://github.com/d3sandoval/openemr/commits/1600-move-away-from-bower).
- Some older versions of jquery (1.3.2 and 1.4.3) and the datatable libraries are now using the unminified scripts because that's what came in the distributable package.

### TODOs (maybe after this PR?)
- integrate with composer. The code is there to use a composer-installed version of NPM but i noticed composer wasn't used in any Dockerfiles. So i've left a few comments and might need some help with this.
- remove the unnecessary after-build steps in build.xml since we're just not copying these files in the first place.

## "Building" the assets
- `npm install` is all it takes to download the versions of the assets from the internet and copy into the `public/assets` directory. See [this commit for an example](https://github.com/d3sandoval/openemr-devops/commit/dc7a35bd5b27872acb8ec6d44509ea01432d6d64). I'll need some help making this build step as seamless as possible... but running this example docker-compose, i was able to test my changes :)

## Some fun things that come out of this
- We no longer commit (most) vendor libraries! And adding new vendor libraries is as easy as `npm install` (and updating the `config.yaml` if you want to use it in the `Header` function)
- Any package is also available in the root `node_modules` directory - opening up the ability to use them all in a modern front-end stack (like React or Vue).
    - Webpack is already hanging around as a dependency of storybook (the front-end docs app) but we can make that an explicit dependency, if we want to start building modular front-end components outside of PHP

---
below added by bradymiller:
Up For Grabs demos for this PR is here:

Alpine is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Nu_-_Up_For_Grabs_Demo
OpenEMR packages produced by this demo are here:
http://ten.openemr.io/b/files/

Ubuntu is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Xi_-_Up_For_Grabs_Demo
OpenEMR packages produced by this demo are here:
http://seven.openemr.io/d/files/

### CODE REVIEW

File: config/config.yaml [hunk 1]
```diff
@@ -11,114 +11,170 @@
 #         link: jquery-ui.min.css filename or null
 #         autoload: true (defaults to false)
 assets:
```
Comment (bradymiller) [inline config/config.yaml]: alphabetical makes sense, but aren't some of these dependent on order.
For example, doesn't jquery need to ones that depend on jquery?

**Comment by bradymiller** [comment]
hi @d3sandoval ,
Just awesome to see this PR.
Would it make sense to go all in on this and remove all packages in vendor?
(then the build step looks like it is just `composer install`, which appears to call npm build also)
-brady

**Comment by d3sandoval** [comment]
@bradymiller I can try that! The only thing was that i noticed none of the Dockerfiles were actually doing a `composer install`. Can you point me to where we do that in our build process?

When I did it on my local machine, I found that a lot of the stuff that's currently committed to vendor is very different from what actually ended up in the vendor directory after `composer` finished.

File: config/config.yaml (hunk 1 above)
Comment (d3sandoval) [reply config/config.yaml]: hrm. good point. I'll have a look at Header.php to see how these things are being loaded. I'd prefer to keep this config in order and use the Header file to manage what loads in what order. Most script managers do this by having a "depends:" key that only loads a script after all of its dependencies have been loaded first.

**Comment by bradymiller** [comment]
Sounds like first step is to get a clean composer install in the codebase. See here for proposed cleanup/reinstall:
https://github.com/openemr/openemr/pull/1768
(I think this is right, but am a little unknowing about correctly setting the php versioncontraint; note if we don't set this, it seems to use current installed php version on the box building it, which is not what we want, though)

**Comment by bradymiller** [comment]
Note the above composer clean up PR is now in the codebase :)

And here is the proposal to support this in the flex dockers (I should be able to test these dockers and get them out tomorrow evening). The goal is that composer install does it all for developers needs(and would be nice for all the npm stuff to happen with that command also); the other lines after that are just clean up and optimization for the final production builds.
https://github.com/openemr/openemr-devops/pull/150

Note that to get your code using this composer build, all you will need to to is completely delete the vendor directory and bring over the composer.json, composer.lock, and build.xml from the main codebase (when you are ready to wipe out the vendor directory and the above flex dockers are ready).

It's going to take some time to get this into all the build pipelines, so could even consider incorporating the css building in this :)
(if did this, rec migrating the final css to a directory to allow gitignore)

**Comment by bradymiller** [comment]
In above flex dockers, note the reason I have the FORCE_NO_BUILD_MODE environment variable is in the case of the dev docker environment where I have 4 flex dockers on 1 host openemr directory. In this case, I set 3 of 4 of the dockers to have FORCE_NO_BUILD_MODE so there is not some bizarre race condition that happens where they are building packages (if no vendor directory exists) at the same time in the same place. Here's the changes in the mega docker-compose.yml to illustrate:
https://github.com/openemr/openemr/commit/1a0c0b06d05a418d30997d59bb8e9fb0d2bc3878

**Comment by bradymiller** [comment]
hi @d3sandoval ,
The new flex dockers are now updated. If the vendor directory is missing, then they will run the compose install stuff.
-brady

**Comment by d3sandoval** [comment]
@bradymiller perfect! Are we expecting that, if we merge this to master, that the 5.0.2 build scripts will have the same composer logic? I'll work with the flex docker for the sake of this PR - but I'm not sure if I'm the best person to edit the 5.0.2 Dockerfile, if it needs editing to support this new installation strategy.

RE: "could even consider incorporating the css building in this"
OK :D This PR will end up bringing that in, assuming nobody is changing the SCSS files while a container is running. Once this is merged, I'll make a separate PR to account for a "development" workflow that watches changes to the SCSS files and updates the compiles CSS.

**Comment by bradymiller** [comment]
hi @d3sandoval ,
I'll be happy to bring this into the Dockerfile for 5.0.2 when this gets into master (it will be the same as flex, but will not need the `if` stuff). By then, we'll have good practice, since I also need to get this going in the demo farm dockers/scripts (both alpine and ubuntu based) (this is why it make take some time to get the pipeline fully worked out so things don't come screeching to a halt when the "build" codebase is introduced).
-brady

**Comment by bradymiller** [comment]
If you do need an interim step here (so things like scripts with the changed header paths do not go stale and start getting merge conflicts), one option is to commit everything you have in public/assets now along with the config.yaml stuff (post build). Then we get testing of all the script path changes and avoid code merge conflicts down the road.
-brady

**Comment by d3sandoval** [comment]
@bradymiller can you please have a look at my suggested docker files? The intent is for this to run for production builds: https://github.com/d3sandoval/openemr-devops/commit/4d8b95b0fea8a368f66f15ca5212f54967757a0a

I've got the build working, installing everything from composer and npm as well as building the css. RTL still needs some work. But I'd like to know if you see anything funky in my suggested build steps.

Thanks!

**Comment by bradymiller** [comment]
hi @d3sandoval ,

Assuming you are ok with a non-existent vendor directory (rather than an empty one), here are PRs for both the openemr-devops (flex and prepping 5.0.2) and demo farm (only will work on alpine so far) respectively:
https://github.com/openemr/openemr-devops/pull/151
https://github.com/openemr/demo_farm_openemr/pull/30

-brady

**Comment by bradymiller** [comment]
hi @d3sandoval ,
Here's the Up For Grabs demo that is running your branch (looks like it is working :) )
https://www.open-emr.org/wiki/index.php/Development_Demo#Nu_-_Up_For_Grabs_Demo

Also, here are the openemr packages that it builds:
http://ten.openemr.io/b/files/

**Comment by bradymiller** [comment]
Took a look in the packages. And looks like may be an issue with the rtl themes:
image

**Comment by bradymiller** [comment]
hi @d3sandoval ,
Note the demo resets daily at midnight (or maybe 1:00 am). If needed, I can also reset it (it literally is just one click of a button for me), so just let me know if needed.

**Comment by bradymiller** [comment]
btw, woohoo on the `Files changed 13,491`!!

**Comment by d3sandoval** [comment]
Thanks @bradymiller ! I've almost the rtl files. I have to change the way their built since they previously relied on the generated css files living in the interface/themes folder.

In addition, some pages still try to get the theme from the old directory.

Once these two issues are resolved, I think this will be ready to merge!

**Comment by d3sandoval** [comment]
I have been struggling to figure out why the pages keep looking for the css files in `/interface/themes/` instead of my new `/public/themes` directory. Any help here is very much appreciated as I feel like i've replaced every (currently used) instance of /interface/themes in the application.

Other than this issue causing most pages to 404 on getting the CSS files, this feels ready to merge. If anyone in @openemr/admins can point me in the right direction on where all of the scripts are trying to pull the css file location from, i would be very grateful!

**Comment by bradymiller** [comment]
hi @d3sandoval ,
In theory, you should only need to change the following paths for main css style paths to work:
https://github.com/openemr/openemr/blob/master/interface/globals.php#L348
https://github.com/openemr/openemr/blob/master/interface/globals.php#L429
https://github.com/openemr/openemr/blob/master/interface/globals.php#L432

**Comment by bradymiller** [comment]
one more here :)
https://github.com/openemr/openemr/blob/master/interface/globals.php#L457

**Comment by bradymiller** [comment]
btw, although unrelated, the line here seems really odd:
https://github.com/openemr/openemr/blob/master/interface/globals.php#L538

Any reason why I shouldn't remove that line?

**Comment by bradymiller** [comment]
Sorry for the trail of messages. But this seems odd because it does not appear that $css_header is being set in the main globals.php code (only $GLOBALS['css_header'] is set).

**Comment by bradymiller** [comment]
Just address the above issue here (I just brought it into the codebase):
https://github.com/openemr/openemr/pull/1787

**Comment by d3sandoval** [comment]
@bradymiller It's working! Thanks for the suggestions. All css files (and their dependencies) now load correctly:
image

Before merging, I would like to ask that someone with a bit more knowledge of common use-cases to a click through of the interface - just to make sure I didn't miss anything.

If nobody finds a problem, it's safe to say that bower has been successfully replaced with npm! More documentation on both package management and using npm with docker in development will come in a future PR... to avoid having to keep this branch up to date ;)

**Comment by bradymiller** [comment]
Great news! I just reset the up for grabs demo. Lets also make sure we have instructions for developers that are using main codebase in the main readme.md .

**Comment by bradymiller** [comment]
@danora24 ,
Getting following errors in the setup log for the up for grabs demo:
```
Generating autoload files

> node-sass@4.9.2 install /var/www/localhost/htdocs/b/openemr/node_modules/node-sass
> node scripts/install.js

Cached binary found at /root/.npm/node-sass/4.9.2/linux_musl-x64-57_binding.node

> uglifyjs-webpack-plugin@0.4.6 postinstall /var/www/localhost/htdocs/b/openemr/node_modules/webpack/node_modules/uglifyjs-webpack-plugin
> node lib/post_install.js


> node-sass@4.9.2 postinstall /var/www/localhost/htdocs/b/openemr/node_modules/node-sass
> node scripts/build.js

Binary found at /var/www/localhost/htdocs/b/openemr/node_modules/node-sass/vendor/linux_musl-x64-57/binding.node
Testing binary
Binary is fine

> openemr-interface@0.2.0 postinstall /var/www/localhost/htdocs/b/openemr
> napa && gulp -i

info cache https://github.com/asyraf9/bootstrap-sidebar/archive/v0.2.2.tar.gz into bootstrap-sidebar
info cache https://github.com/jquery/jquery/archive/1.2.2.tar.gz into jquery-1-2-2
info cache https://github.com/jquery/jquery/archive/1.3.2.tar.gz into jquery-1-3-2
info cache https://github.com/jquery/jquery/archive/1.4.3.tar.gz into jquery-1-4-3
info cache https://github.com/jquery/jquery/archive/1.7.2.tar.gz into jquery-1-7-2
info cache https://github.com/jquery/jquery/archive/1.9.1.tar.gz into jquery-1-9-1
info cache https://github.com/jquery/jquery/archive/1.10.2.tar.gz into jquery-1-10-2
info cache https://github.com/jquery/jquery/archive/1.11.3.tar.gz into jquery-1-11-3
info cache https://github.com/PawelDecowski/jquery-creditcardvalidator/archive/v1.1.0.tar.gz into jquery-creditcardvalidator
info cache https://github.com/eduardomb/jquery-panelslider/archive/0.1.1.tar.gz into jquery-panelslider
info cache https://jqueryui.com/resources/download/jquery-ui-1.10.4.zip into jquery-ui-1-10-4
info cache https://jqueryui.com/resources/download/jquery-ui-1.11.4.zip into jquery-ui-1-11-4
info cache https://jqueryui.com/resources/download/jquery-ui-1.12.1.zip into jquery-ui
info cache https://jqueryui.com/resources/download/jquery-ui-themes-1.10.4.zip into jquery-ui-themes-1-10-4
info cache https://jqueryui.com/resources/download/jquery-ui-themes-1.11.4.zip into jquery-ui-themes-1-11-4
info cache https://jqueryui.com/resources/download/jquery-ui-themes-1.12.1.zip into jquery-ui-themes
info cache https://github.com/jquery-validation/jquery-validation/releases/download/1.13.1/jquery-validation-1.13.1.zip into jquery-validation
info cache https://github.com/EastDesire/jscolor/archive/v1.4.5.tar.gz into jscolor-1-4-5
info cache https://github.com/literallycanvas/literallycanvas/archive/v0.4.13.tar.gz into literallycanvas
info cache https://github.com/pure-css/pure/releases/download/v0.5.0/pure-0.5.0.tar.gz into pure
info cache https://github.com/facebook/react/releases/download/v15.1.0/react-15.1.0.zip into react-15-1-0
info cache https://github.com/pHAlkaline/summernote-plugins into summernote-nugget
[22:29:11] Using gulpfile /var/www/localhost/htdocs/b/openemr/gulpfile.js
[22:29:11] Starting 'install'...
[22:29:12] Finished 'install' after 60 ms
[22:29:12] Starting 'default'...
[22:29:12] Finished 'default' after 34 Î¼s
npm WARN openemr-interface@0.2.0 No repository field.
npm WARN openemr-interface@0.2.0 license should be a valid SPDX license expression
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

added 1764 packages in 39.729s

> openemr-interface@0.2.0 build /var/www/localhost/htdocs/b/openemr
> gulp -b

[22:29:19] Using gulpfile /var/www/localhost/htdocs/b/openemr/gulpfile.js
[22:29:19] Starting 'default'...
[22:29:19] Starting 'ingest'...
[22:29:19] Finished 'ingest' after 148 Î¼s
[22:29:19] Starting 'clean'...
[22:29:19] Finished 'clean' after 2.88 ms
[22:29:19] Starting 'styles:style_uni'...
[22:29:19] Finished 'styles:style_uni' after 24 ms
[22:29:19] Starting 'styles:style_color'...
[22:29:19] Finished 'styles:style_color' after 2.97 ms
[22:29:19] Starting 'styles:rtl'...
[22:29:20] Finished 'styles:rtl' after 39 ms
[22:29:20] Starting 'styles'...
[22:29:20] Finished 'styles' after 45 Î¼s
[22:29:20] Starting 'sync'...
[22:29:20] Finished 'sync' after 2.74 ms
[22:29:20] Starting 'style_list'...
[22:29:20] 'style_list' errored after 498 Î¼s
[22:29:20] Error: ENOENT: no such file or directory, open '.storybook/themeOptions.json'
    at Object.fs.openSync (fs.js:646:18)
    at Object.fs.writeFileSync (fs.js:1299:33)
    at Gulp.<anonymous> (/var/www/localhost/htdocs/b/openemr/gulpfile.js:195:8)
    at module.exports (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:273:3)
    at Gulp.Orchestrator._runStep (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:214:10)
    at Gulp.Orchestrator.start (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:134:8)
    at runNextSet (/var/www/localhost/htdocs/b/openemr/node_modules/run-sequence/index.js:124:15)
    at Gulp.onTaskEnd (/var/www/localhost/htdocs/b/openemr/node_modules/run-sequence/index.js:102:4)
    at emitOne (events.js:121:20)
[22:29:20] 'default' errored after 80 ms
[22:29:20] [31mError[39m in plugin "[36mrun-sequence(style_list)[39m"
Message:
    ENOENT: no such file or directory, open '.storybook/themeOptions.json'
Details:
    errno: -2
    code: ENOENT
    syscall: open
    path: .storybook/themeOptions.json

Stack:
Error: ENOENT: no such file or directory, open '.storybook/themeOptions.json'
    at Object.fs.openSync (fs.js:646:18)
    at Object.fs.writeFileSync (fs.js:1299:33)
    at Gulp.<anonymous> (/var/www/localhost/htdocs/b/openemr/gulpfile.js:195:8)
    at module.exports (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:273:3)
    at Gulp.Orchestrator._runStep (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:214:10)
    at Gulp.Orchestrator.start (/var/www/localhost/htdocs/b/openemr/node_modules/orchestrator/index.js:134:8)
    at runNextSet (/var/www/localhost/htdocs/b/openemr/node_modules/run-sequence/index.js:124:15)
    at Gulp.onTaskEnd (/var/www/localhost/htdocs/b/openemr/node_modules/run-sequence/index.js:102:4)
    at emitOne (events.js:121:20)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! openemr-interface@0.2.0 build: `gulp -b`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the openemr-interface@0.2.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2018-08-12T22_29_29_122Z-debug.log
```

**Comment by bradymiller** [comment]
btw, here are the commands that the up for grabs demo runs:
https://github.com/openemr/demo_farm_openemr/blob/master/demo_build.sh#L473-L494

**Comment by d3sandoval** [comment]
That's really weird. `.storybook/themeOptions.json` is committed. Could it be a permissions issue? What user is running the npm script?

**Comment by d3sandoval** [comment]
It's not necessary to build that json file every time. so we can just leave it out of the "prod" build, for now. Pushed a new commit to see if that fixes things.

**Comment by bradymiller** [comment]
hi @d3sandoval ,

Up for grabs demo is still not working (can't log in, although the setup log looks fine).

I tried this locally, and was able to login, but getting lots of errors logging in the js console (also, datetimepicker does not work), which all appear to be related to the incorrect ordering in the config.yaml file; rec. ordering those as they should be ordered in the scripts rather than alphabetical.

-brady

**Comment by bradymiller** [comment]
Also will be helpful to migrate the .docs/.out/ .gitignore into the main .gitignore (easier to track when all in one main .gitignore).

**Comment by d3sandoval** [comment]
@bradymiller new commits should fix those issues. thanks for the heads up on the config ordering :) I don't see any errors on my machine now:
image

Anything else I can do to get this working in the demo environment? Happy to help as much as I can, although the build system still feels like a bit of a black box.

**Comment by d3sandoval** [comment]
@bradymiller I just tried the demo and it's giving me the following error when i'm trying to login: Invalid username or password

Is there any way to know why this issue is occuring? It seems to be only a problem with the demo environment as i can login just fine on a local docker.

**Comment by bradymiller** [comment]
@d3sandoval ,
Looks like I ordered things wrong on the demo farm script :) ; I'll plan to fix it today.
-brady

**Comment by bradymiller** [comment]
@d3sandoval ,

Got the Alpine demo working. Now I gotta get the ubuntu demos working:

Alpine is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Nu_-_Up_For_Grabs_Demo
OpenEMR packages produced by this demo are here:
http://ten.openemr.io/b/files/

Ubuntu is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Xi_-_Up_For_Grabs_Demo
OpenEMR packages produced by this demo are here:
http://seven.openemr.io/d/files/

-brady

**Comment by d3sandoval** [comment]
@bradymiller this is great! I'm able to login just fine to both instances! I'm working on documentation now so it's clear how to build things locally and within the docker container. I can put that in a separate PR, if you're thinking about merging this soon.

**Comment by bradymiller** [comment]
This is getting very close and agree can release main documentation in a later PR. Just want to confirm that these are what developers need to do to build their local openemr? :
```
composer install
npm install
npm run build
```

**Comment by bradymiller** [comment]
IN the config.yaml, does `tabs-theme` have the correct path??

**Comment by bradymiller** [comment]
The rtl themes does not look right either:
what rtl should look like:
image
what it looks like in this PR:
image

**Comment by d3sandoval** [comment]
@bradymiller I'll take a look at those issues. thanks for finding them.

As far as docs go, I've created a branch that adds documentation and a new (safer) docker-compose file. You can check it out here: https://github.com/d3sandoval/openemr/blob/new-build-docs/contrib/util/docker/docker-compose.yml
Most changes are in the Contributing Doc: https://github.com/d3sandoval/openemr/blob/new-build-docs/CONTRIBUTING.md

This change requires that we ignore errors from rsync and chown, an example of that can be found here: https://github.com/d3sandoval/openemr-devops/commit/e6fabd4d492f87a05d665c6f23e076b57709bb3d

**Comment by d3sandoval** [comment]
RE: what developers need to do to build their local openem
They only need to do `npm install` because `npm run build` will run automatically after installation.

**Comment by d3sandoval** [comment]
Fixed the tab issues with the latest commit:
image

**Comment by d3sandoval** [comment]
BTW. making that fix was so incredibly easy with the new docker-compose file. No waiting for files to copy or accidentally seeing stale files in my filesystem. Just a clean docker instance that I could hack on and see my changes :)

**Comment by bradymiller** [comment]
I just blew up the demo farm and fixing it now :)
(removing the `npm run build` line from the build to ensure it works)

btw, shouldn't %assets_static_relative% be used in the tabs-theme path?

**Comment by d3sandoval** [comment]
OK I lied. Removing `npm run build` doesn't work... i confused the `postinstall` and `build` scripts.

I should have checked first...

**Comment by d3sandoval** [comment]
> btw, shouldn't %assets_static_relative% be used in the tabs-theme path?

I don't think so. `%assets_static_relative%` points to `/public/assets` while the tabs theme is located in `/public/themes`

**Comment by bradymiller** [comment]
you are correct. (I haven't had enough coffee...) :)

**Comment by bradymiller** [comment]
ok, I think this is ready for prime time. going to bring it in now :)

**Comment by kofiav** [comment]
:clap:

**Pull request merged**